### PR TITLE
feat(shape): Added new Shape APIs

### LINKED ISSUE #3175: Add Shape system APIs
Systematize shape system to allow components to theme based on shape category. Create SASS base shape mixin which can be used by all other components. New shape variables to allow global overrides and shape functions to reuse in SASS mixin.

### PR DESCRIPTION
**Pull request by abhiomkar**
Previously on PR #3490

Fixes #3175 

BREAKING CHANGE: The previous contents of the mdc-shape package have been removed and replaced with mixins implementing the Shape system. This system implements only rounded corners to provide a straightforward CSS-only solution.

### CODE REVIEW

File: packages/mdc-shape/_functions.scss [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+// $radius can be a single value or list of 4.
+//
+// Examples:
+//
+// mdc-shape-prop-value(small) => 4px
+//
+@function mdc-shape-prop-value($radius) {
+  @if type-of($radius) == "list" {
+    @if length($radius) > 4 {
+      @error "Invalid radius: '#{$radius}' radius is not supported";
```
Comment (kfranqueiro) [inline packages/mdc-shape/_functions.scss]: This error message could be more specific for what it's being raised for

File: packages/mdc-shape/_variables.scss [hunk 2]
```diff
@@ -20,39 +20,14 @@
 // THE SOFTWARE.
 //
 
-.mdc-shape-container {
-  display: inline-block;
-  position: relative;
-  overflow: hidden; // Prevent overflowed borders from encroaching on surrounding elements
-}
+// Shape categories
```
Comment (kfranqueiro) [inline packages/mdc-shape/_variables.scss]: Are these really the correct defaults? Will all 3 of these values end up being used OOTB? It seems like we might end up with more exceptions than inclusions with regard to these "buckets".

File: packages/mdc-shape/_functions.scss [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+@import './variables';
+
+//
```
Comment (kfranqueiro) [inline packages/mdc-shape/_functions.scss]: Any public functions here should be documented in a table in the readme

File: packages/mdc-shape/_variables.scss (hunk 2 above)
Comment (abhiomkar) [reply packages/mdc-shape/_variables.scss]: Yes, these are updated baseline values for shape categories.

File: packages/mdc-shape/_functions.scss (hunk 1 above)
Comment (abhiomkar) [reply packages/mdc-shape/_functions.scss]: Done.

File: packages/mdc-shape/_functions.scss (hunk 3 above)
Comment (abhiomkar) [reply packages/mdc-shape/_functions.scss]: Done.

File: packages/mdc-shape/README.md [hunk 4]
```diff
@@ -36,74 +35,50 @@ applying angled corners to unelevated surfaces.
 npm install @material/shape
 ```
 
-## Basic Usage
+## Usage
```
Comment (kfranqueiro) [inline packages/mdc-shape/README.md]: Discussed offline: let's bring this back in line with the readme standards/template.

File: packages/mdc-shape/_functions.scss [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+// Returns $radius value of shape category - `large`, `medium` or `small`.
+// Otherwise, it returns the $radius itself if valid.
+// $radius can be a single value or list of 4.
+//
+// Examples:
+//
+// mdc-shape-prop-value(small) => 4px
+//
+@function mdc-shape-prop-value($radius) {
```
Comment (kfranqueiro) [inline packages/mdc-shape/_functions.scss]: Do we actually have any use case where this needs to be used outside this package? I don't see it used in the other 2 PRs so far. Should we make it private until a need for it arises?

File: packages/mdc-shape/README.md [hunk 6]
```diff
@@ -38,72 +37,72 @@ npm install @material/shape
...
+### Sass Functions
 
-Outlined angled corners involve the same markup and styles/imports as above, with the addition of including a mixin for
-outline:
+Function | Description
+--- | ---
+`mdc-shape-flip-radius($radius)` | Function that flips the radius in RTL context. $radius is list of corner values it can be length of 4, 3 or 2.
+`mdc-shape-resolve-percentage-radius($component-height, $radius)` | Function that calculates the absolute radius value based on its component height. Use this for fixed height components only.
+`mdc-shape-prop-value($radius)` | Returns radius value of shape category - `large`, `medium` or `small`. Otherwise, it returns the `$radius` itself if valid. `$radius` can be a single value or list of 4.
+`mdc-shape-mask-top-radius($radius)` | Accepts radius number or list of 2 radius values and returns 4 value list with top-left value, top-right value, bottom-right 0, bottom-left 0. Throws error when length of radius is more than 2.
```
[the remark this answers is not in the document]
Comment (kfranqueiro) [reply packages/mdc-shape/README.md]: I just found this defined in #3552. You might want to move the function to this PR, or move these docs to that PR instead.

File: packages/mdc-shape/_functions.scss [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    }
+    @return $radius-value;
+  } @else {
+    @return mdc-shape-resolve-percentage-for-corner_($component-height, $radius);
+  }
+}
+
+@function mdc-shape-resolve-percentage-for-corner_($component-height, $radius) {
+  @if type-of($radius) == "number" and unit($radius) == "%" {
+    // Converts the percentage to number without unit. Example: 50% => 50.
```
[the remark this answers is not in the document]
Comment (abhiomkar) [reply packages/mdc-shape/_functions.scss]: I feel documenting what this accomplishes is important than explaining implementation details. Developer would figure out how this works once he know what it does.

File: packages/mdc-shape/_functions.scss (hunk 5 above)
Comment (abhiomkar) [reply packages/mdc-shape/_functions.scss]: This is being used in mixins.scss which needs this function to be public. Assuming _ suffix is used for file scoped.

File: packages/mdc-shape/_functions.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (kfranqueiro) [reply packages/mdc-shape/_functions.scss]: In general that's probably true, but this is basically taking advantage of specific Sass behavior to workaround what basically amounts to a missing API. I was able to surmise what it accomplished almost immediately, but it looked mysterious and took me a minute to rationalized why it worked. My request was to save others that minute.

File: packages/mdc-shape/_functions.scss (hunk 5 above)
Comment (kfranqueiro) [reply packages/mdc-shape/_functions.scss]: We currently have cases where we use trailing underscore to indicate that something can be used within the package but shouldn't be used outside of it. See mdc-typography/functions.scss for examples. (Granted, this is not how private members will work in Sass modules...)

I guess this one can go either way, I just figured exposing less is better if it's unlikely to be needed by end-user developers.

Comment (abhiomkar) [reply packages/mdc-shape/_functions.scss]: Done.

File: packages/mdc-shape/README.md [hunk 8]
```diff
@@ -38,72 +37,72 @@ npm install @material/shape
...
 
-.my-shape-container {
-  @include mdc-shape-angled-corner(#fff, 10px);
-}
-```
+### Sass Mixins
+
+Mixin | Description
+--- | ---
+`mdc-shape-radius($radius, $rtl-reflexive)` | Shape API used by all other components to apply radius to appropriate corners. `$radius` can be single value or list of 4 radius corner values. Set `$rtl-reflexive` to true to flip the radius in RTL case, `false` by default.
```
[the remark this answers is not in the document]
Comment (kfranqueiro) [reply packages/mdc-shape/README.md]: It looks like this wasn't updated?

File: packages/mdc-shape/_functions.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (abhiomkar) [reply packages/mdc-shape/_functions.scss]: Moved this to a private function so it won't clobber the code.

File: packages/mdc-shape/README.md [hunk 9]
```diff
@@ -38,72 +37,71 @@ npm install @material/shape
...
-`mdc-shape-angled-corner-outline($outline-width, $outline-color[, $outline-style])` | Applies outline styles to angled corners. `$outline-style` defaults to `solid`.
+Where, `$radius` is absolute value only.
+
+#### Shapes for components on specific corners
 
-> **Note:** When mentioned above, "background color" specifically refers to the color of the background behind the surface (_not_ the fill color of the surface). These mixins operate by masking the corners of the surface to match the background.
+Styles for applying shapes for specific corners such as drawer looks like this:
+
+```scss
+@include mdc-shape-radius((0, $radius, $radius, 0), $rtl-reflexive: true);
```
Comment (kfranqueiro) [inline packages/mdc-shape/README.md]: Can this be re-written as `@include mdc-shape-radius(0 $radius $radius 0, $rtl-reflexive: true);` ?

File: packages/mdc-shape/README.md [hunk 10]
```diff
@@ -38,72 +37,71 @@ npm install @material/shape
...
+> Use `mdc-shape-resolve-percentage-radius` sass function to resolve percentage unit value to absolute radius value.
 
-### Outlined Angled Corners
+### Sass Functions
 
-Outlined angled corners involve the same markup and styles/imports as above, with the addition of including a mixin for
-outline:
+Function | Description
+--- | ---
+`mdc-shape-flip-radius($radius)` | Flips the radius values in RTL context. $radius is list of 2-4 corner values.
```
Comment (kfranqueiro) [inline packages/mdc-shape/README.md]: Add backticks around `$radius`

File: packages/mdc-shape/_functions.scss [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+      $radius-value: append($radius-value, mdc-shape-prop-corner-value_($corner));
+    }
+
+    @return $radius-value;
+  } @else {
+    @return mdc-shape-prop-corner-value_($radius);
+  }
+}
+
+//
```
Comment (kfranqueiro) [inline packages/mdc-shape/_functions.scss]: Move this up above internal functions?

**Pull request merged**